File: examples/Composers.py
#!/usr/bin/env python

# ------------
# Composers.py
# ------------

import operator

def bind2nd (bf, y) :
    return lambda x : bf(x, y)

def compose (f, g) :
    return lambda x : f(g(x))

print "Composers.py"

f = compose(bind2nd(operator.add, 1), bind2nd(operator.mul, 3))
assert f(5) == 16

print "Done."

File: examples/Composers.hs
-- ------------
-- Composers.hs
-- ------------

{-
(>>)     :: IO a          -> IO b      -> IO b -- 'then' operator
assert   :: Bool          -> a         -> a
flip     :: (a -> b -> c) -> b    -> a -> c
putStrLn :: String                     -> IO ()
return   :: a                          -> IO a
-}

import Control.Exception (assert)

main :: IO ()
main =
    putStrLn "Composers.hs" >>

    let f :: Int -> Int
        f = (+) 1 . (*) 3
    in assert (f 5 == 16) return () >>

    putStrLn "Done."
